### ISSUE TITLE
abstract and String concat

### ISSUE THREAD

**Issue by andyli** [HaxeFoundation/haxe#8627]
```haxe
abstract Decimal(Float) from Float to Float {
    @:from static function fromString(value : String) : Decimal
        return Std.parseFloat(value);

    @:to public function toString() : String
        return Std.string(this);

    @:op(A+B) @:commutative
    function add(that : Decimal) : Decimal
        return this + (that:Float);
}

class Test {
    static function main() {
        var d:Decimal = 123.4;
        trace("str: " + d);            // Test.hx:16: NaN, but I expect it to be "str: 123.4"
        trace("str: " + d.toString()); // Test.hx:17: str: 123.4, correct
    }
}
```

Tested with 4.0.0-rc.3+348192fff.
With Haxe 3.4, both traces are `str: 123.4`.

From the output of Haxe 4, I can see that it's trying to convert `"str: "` to `Decimal` and do a `+` operation with `d`:
```js
Test.main = function() {
	var d = 123.4;
	console.log("Test.hx:16:",_$Test_Decimal_$Impl_$.toString(_$Test_Decimal_$Impl_$.add(d,_$Test_Decimal_$Impl_$.fromString("str: "))));
	console.log("Test.hx:17:","str: " + _$Test_Decimal_$Impl_$.toString(d));
};
```

**Comment by RealyUniqueName** [comment]
Idk if we should change the behavior here.
While you expect `toString` to be applied instead of `add`, somebody else might expect current behavior.
Ideally compiler would emit a warning about ambiguous operation.

**Comment by Ilir-Liburn** [comment]
Very good example. I think problem comes from the compiler, not design. Compiler is converting "str:" to Decimal. Why? It is clearly stated that @:op(A+B) is taking Decimal parameter, e.g. d + d should trigger @:op(A+B) for example, not "str:" + d. Correct me if I'm wrong.

**Comment by Aurel300** [comment]
@Ilir-Liburn The compiler is converting `"str:"` to decimal because there is a `@:from` function which converts `String` to `Decimal`. `("str:" : Decimal) + d` then naturally uses the `add` operator.

**Comment by Ilir-Liburn** [comment]
("str:":Decimal) + d should trigger @:op(A+B) then?

**Comment by Aurel300** [comment]
Yes, the `+` at that point is just the addition of two `Decimal` values.

**Comment by Ilir-Liburn** [comment]
Yes, but compiler is assuming here that user want to use @:op(A+B), because of the implicit cast. I think in such case, compiler should throw an error forcing user to be explicit. Or warn, but I would prefer error here.

**Comment by Aurel300** [comment]
But that's the point of the implicit cast…

**Comment by Ilir-Liburn** [comment]
But again, it happens because @:op(A+B) is @:commutative. I will say this only once, to not repeat myself from time to time. Worst thing for me is when compiler assumes something for me. If there are two possible outcomes, compiler should at least worn about that.

**Comment by Aurel300** [comment]
We can simplify the example to not use `@:commutative`:

```haxe
abstract Decimal(Float) from Float to Float {
  @:from static function fromString(value:String):Decimal
    return Std.parseFloat(value);

  @:to public function toString():String
    return Std.string(this);

  @:op(A+B) function add(that:Decimal):Decimal
    return this + (that:Float);
}

class Main {
  static function main() {
    var d:Decimal = 123.4;
    trace(d.toString() + "str"); // 123.4str
    trace(d + "str"); // NaN or runtime error
  }
}
```

**Comment by Ilir-Liburn** [comment]
With @:commutative we can do only this

```haxe
trace("str: " + (d:Float));
```
But how will newcomer to Haxe know what is going on under the hood? Should newcomer know about underlying type of some core abstracts for example? Etc, etc.

**Comment by andyli** [comment]
I understand how the expression can be ambiguous in semantic and agree with what you discussed. I want to point out 2 things:

1. The behavior of the posted example changed between Haxe 3 and 4. If we decide the newer behavior is better, we need to justify it with good reasons since it's breaking.

2. The actual code in which I discovered the issue was using string interpolation instead of mere concatenation. i.e. It was `trace('str:  $d');`, which would give a stronger impression of converting `d` to String is the correct behavior.

**Comment by RealyUniqueName** [comment]
I agree this may be considered a regression. @Simn ?

**Comment by Simn** [comment]
I find it very hard to reason about the combination of implicit casts and a type-based selection algorithm like operator overloads. So I'd default to not changing the behavior for now, unless somebody has a clear vision for how this should behave.

In either case, I think we should align operator overloads (including `@:arrayAccess`) with static extension resolution, which is another type-based selection algorithm.

Related: #8427

**Comment by Ilir-Liburn** [comment]
This is how I see it: compiler doesn't respect (builtin) operator for String, e.g. lhs ("str:") should have priority here, not rhs (d). Anyway, it is strange to see extern class String mentioning concatenation using the '+' operator, yet there is no one defined for the String class. Maybe it's a time to make a String abstract type? (was afraid to speak this out loud)

**Comment by RealyUniqueName** [comment]
Moving to 4.1 because it will cause regressions for sure.
